[WIP] [Servo] Refactoring servo

ibrahiminfinite opened the pull request:
### Description

This is an effort supported by GSoC 2023 to refactor the MoveIt Servo package to improve its usability and performance.

The joint jog and twist based demos for the C++ API are currently working.
Please feel free to test it out and provide your feedback.

TODO :
- [ ] C++ joint jog API + demo
- [ ] C++ twist API + demo
- [ ] C++ pose API + demo
- [ ] Tests for C++ APIs
- [ ] ROS start, stop services
- [ ] ROS input type switching service
- [ ] ROS joint jog API + demo
- [ ] ROS twist API + demo
- [ ] ROS pose API + demo
- [ ] Tests for ROS APIs
- [ ] Keyboard teleop demo
- [ ] Joystick teleop demo
- [ ] Update moveit_tutorials

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,101 @@
+  auto servo_params = servo_param_listener->get_params();
+
+  // The publisher to send trajectory message to the robot controller.
+  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr trajectory_outgoing_cmd_pub;
+  trajectory_outgoing_cmd_pub = servo_node->create_publisher<trajectory_msgs::msg::JointTrajectory>(
+      servo_params.command_out_topic, rclcpp::SystemDefaultsQoS());
```
```suggestion
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr trajectory_outgoing_cmd_pub = servo_node->create_publisher<trajectory_msgs::msg::JointTrajectory>(
      servo_params.command_out_topic, rclcpp::SystemDefaultsQoS());
```

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,101 @@
+  // Set the command type for servo.
+  servo.incomingCommandType(CommandType::JOINT_JOG);
+
+  RCLCPP_INFO_STREAM(LOGGER, servo.getStatusMessage());
```
Seems like this line doesn't do anything?

```suggestion
```

ibrahiminfinite replied:
This is here just for convenience to print the status before beginning the demo.

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,101 @@
+  {
+    // Move only the 7th joint
+    moveit_servo::JointJog vec(7);
+    vec << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0;
```
```suggestion
    vec << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0;  // rad/s
```

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,101 @@
+ *      Author    : V Mohammed Ibrahim
+ */
+
+#include <iostream>
+#include <chrono>
+#include <rclcpp/rclcpp.hpp>
+#include <tf2_eigen/tf2_eigen.hpp>
+#include <moveit_servo/servo.hpp>
+#include <moveit_servo/utils.hpp>
```
I believe these are unused but please double-check.

```suggestion
#include <chrono>
#include <rclcpp/rclcpp.hpp>
#include <moveit_servo/servo.hpp>
#include <moveit_servo/utils.hpp>
```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,122 @@
+   * @param carteisan_position_delta The change in cartesian position.
+   * @return The required joing angle deltas.
+   */
+  Eigen::VectorXd detlaFromIkSolver(const Eigen::VectorXd& cartesian_position_delta);
```
```suggestion
  Eigen::VectorXd deltaFromIkSolver(const Eigen::VectorXd& cartesian_position_delta);
```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,122 @@
+
+  kinematics::KinematicsBaseConstPtr ik_solver_ = nullptr;
+
+  uint64_t controller_period_;
+  std::map<std::string, control_toolbox::Pid> controllers_;
```
Check if these are used. I don't see why they would be.
```suggestion
```

ibrahiminfinite replied:
Removed these as well, since the last update moves away from PID based control.

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,122 @@
+  uint64_t controller_period_;
+  std::map<std::string, control_toolbox::Pid> controllers_;
+};
+
```
```suggestion
```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,122 @@
+/*      Title     : command_processor.hpp
+ *      Project   : moveit_servo
+ *      Created   : 04/06/2023
+ *      Author    : Brian O'Neil, Andy Zelenak, Blake Anderson, V Mohammed Ibrahim
```
Please add a Description field. Something like:

```suggestion
 *      Author    : Brian O'Neil, Andy Zelenak, Blake Anderson, V Mohammed Ibrahim
 *      Description    : todo
```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,122 @@
+
+#pragma once
+
+#include <variant>
+#include <control_toolbox/pid.hpp>
+
```
Seem to be unused. Please double-check your includes.

```suggestion
#include <variant>
#include <control_toolbox/pid.hpp>

```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/datatypes.hpp:
```diff
@@ -36,22 +31,52 @@
+
+enum class CommandType
+{
+  JOINT_JOG = 0,
+  TWIST,
+  POSE
```
Might as well initialize all of them

```suggestion
  JOINT_JOG = 0,
  TWIST = 1,
  POSE = 2
```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:
```diff
@@ -0,0 +1,171 @@
+
+  const rclcpp::Node::SharedPtr node_;
+  std::unique_ptr<CommandProcessor> command_processor_;
+  std::atomic<CommandType> incoming_command_type_;
```
Is this accurate?

```suggestion
  // This needs to be threadsafe so it can be updated in realtime with Dynamic Reconfigure
  std::atomic<CommandType> incoming_command_type_;
```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:
```diff
@@ -0,0 +1,171 @@
+  rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr collision_velocity_scale_sub_;
+
+  pluginlib::UniquePtr<online_signal_smoothing::SmoothingBaseClass> smoother_;
+  pluginlib::ClassLoader<online_signal_smoothing::SmoothingBaseClass> smoothing_loader_;
```
It seems like `smoothing_loader_` could be a local variable, it doesn't need to be a class member.

```suggestion
  pluginlib::ClassLoader<online_signal_smoothing::SmoothingBaseClass> smoothing_loader_;
```

ibrahiminfinite replied:
It is now local

tylerjw commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,105 @@
+  rclcpp::init(argc, argv);
+
+  // The servo object expects to get a ROS node.
+  auto servo_node = std::make_shared<rclcpp::Node>("servo_node");
```
mabye this should just be "moveit_servo", we know this is a node from context.

ibrahiminfinite replied:
Oh , that was a poor naming choice for the variable as well.
Since this is not actually a `Servo` node.

sea-bass commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -89,29 +89,30 @@ servo:
-      default_value: 0.5,
-      description: "Max joint angular/linear velocity. Only used for joint commands on joint_command_in_topic."
-    }
+  linear_scale: {
+    type: double,
+    default_value: 0.4,
+    description: "Max linear velocity. Unit is [m/s]. Only used for Cartesian commands."
+  }
+
+  rotational_scale: {
+    type: double,
+    default_value: 0.8,
+    description: "Max angular velocity. Unit is [rad/s]. Only used for Cartesian commands"
+  }
+
+  joint_scale: {
+    type: double,
+    default_value: 0.5,
+    description: "Max joint angular/linear velocity. Only used for joint commands on joint_command_in_topic"
+  }
```
If you merge/rebase on main, these params should retain the original structure

ibrahiminfinite replied:
These are unused now since support for unitless command is not added yet, so I removed them in the last update.
I am not yet sure on how those should be handled. It does not make sense for servo to directly receive a "unitless" command. 
This should probably be handled by a utility function before the command is passed to servo.

sea-bass commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -239,7 +240,7 @@ servo:
                   Collision checking begins slowing down when nearer than a specified distance.",
     validation: {
-      gt_eq<>: 0.0
+      gt_eq<>: 10.0
```
I feel like for some applications a min of 10 Hz might be too high, I think greater than zero is ok

sea-bass commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,105 @@
+/*******************************************************************************
```
These `demo` files as they are would actually serve as good integration tests. Basically do something like:

* Start node
* Record the joint states
* Jog the 7th joint for a while
* Check that the final joint states are different

Unsure if the timing will be accurate enough to actually check the value of the final joint state (probably not), but having it move at least validates the interface

ibrahiminfinite replied:
That is a good idea.
Maybe we can do an Eigen `isApprox` check with a slightly bigger tolerance for where we expect it to be.

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/collision_monitor.hpp:
```diff
@@ -30,24 +30,52 @@
+  // Variables
+  std::atomic<bool> stop_requested_;
+  std::thread monitor_thread_;
+
+  std::atomic<double>& collision_velocity_scale_;
+  const servo::Params& servo_params_;
 
-  rclcpp::NodeOptions options;
+  collision_detection::CollisionRequest self_collision_request_;
+  collision_detection::CollisionResult self_collision_result_;
 
-  auto servo_node = std::make_shared<moveit_servo::ServoNode>(options);
+  collision_detection::CollisionRequest scene_collision_request_;
+  collision_detection::CollisionResult scene_collision_result_;
 
-  rclcpp::spin(servo_node->get_node_base_interface());
+  moveit::core::RobotStatePtr robot_state_;
+  const planning_scene_monitor::PlanningSceneMonitorPtr planning_scene_monitor_;
```
Add comments

ibrahiminfinite replied:
Done

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/datatypes.hpp:
```diff
@@ -36,22 +31,54 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *******************************************************************************/
 
+/*      Title       : datatypes.hpp
```
This file also needs more comments

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/utils.hpp:
```diff
@@ -0,0 +1,152 @@
+/*******************************************************************************
```
mmmm functional interface! Awesome to see

sea-bass commented on moveit_ros/moveit_servo/src/collision_monitor.cpp:
```diff
@@ -0,0 +1,152 @@
+  {
+    if (servo_params_.check_collisions)
+    {
+      self_collision_scale = scene_collision_scale = 1.0;
```
Don't we want these to always be updated to 1.0 even when `servo_params_.check_collisions` is false?

May want to put these outside the `if` above

ibrahiminfinite replied:
Good point.
Have made the change.

sea-bass commented on moveit_ros/moveit_servo/src/command_processor.cpp:
```diff
@@ -0,0 +1,270 @@
+    }
+    else
+    {
+      // Robot does not have an IK solver, use inverse Jacobian to compute IK.
+      Eigen::MatrixXd jacobian = robot_state_->getJacobian(joint_model_group_);
+      Eigen::JacobiSVD<Eigen::MatrixXd> svd =
+          Eigen::JacobiSVD<Eigen::MatrixXd>(jacobian, Eigen::ComputeThinU | Eigen::ComputeThinV);
+      Eigen::MatrixXd matrix_s = svd.singularValues().asDiagonal();
+      Eigen::MatrixXd pseudo_inverse = svd.matrixV() * matrix_s.inverse() * svd.matrixU().transpose();
+
+      joint_position_delta = pseudo_inverse * cartesian_position_delta;
```
Could put this in the utilities file so this looks even cleaner

ibrahiminfinite replied:
I combined this together with the IK solver and made it into a single method called `jointDeltaFromIK`.
Looks much cleaner overall.

sea-bass commented on moveit_ros/moveit_servo/src/command_processor.cpp:
```diff
@@ -0,0 +1,270 @@
+  }
+}
+
+Eigen::VectorXd CommandProcessor::deltaFromIkSolver(const Eigen::VectorXd& cartesian_position_delta)
```
for consistency call this `jointDeltaFromIkSolver`

sea-bass commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,105 @@
+
+/*      Title     : joint_jog_demo.cpp
+ *      Project   : moveit_servo
+ *      Created   : 27/05/2023
```
really nitpicky, but these comments were generated in the US date format which is Month/Day/Year, so for consistency may be worth switching them around. Sorry...

ibrahiminfinite replied:
haha!
Did not notice that, have fixed it for all the files.

sea-bass commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,421 @@
+
+  const CommandType incoming_type = incomingCommandType();
+
+  if (incoming_type == CommandType::JOINT_JOG && command.index() == static_cast<size_t>(incoming_type))
```
Since the `command.index()` check is common to all of the cases, you could pull this out one level above, so

```
if (command.index() == static_cast<size_t>(incoming_type))
{
  if (incoming_type == CommandType::JOINT_JOG)
  {
    ...
```

ibrahiminfinite replied:
Good catch!
This helped make the error message more specific as well.

sea-bass commented on moveit_ros/moveit_servo/src/utils.cpp:
```diff
@@ -0,0 +1,319 @@
+                             double scaling_override)
+{
+  // If override value is close to zero, user is not overriding the scaling
+  if (scaling_override < 0.01)
```
Can we make this 0.01 a constant that is more visible, rather than hard-coded? I've had to explain this one to a lot of people haha

ibrahiminfinite replied:
```c++
namespace
{
// The threshold above which `override_velocity_scaling_factor` will be used instead of computing the scaling from joint bounds.
const double SCALING_OVERRIDE_THRESHOLD = 0.01;
}  // namespace

```

This should be enough ?

sea-bass commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp:
```diff
@@ -0,0 +1,148 @@
+  // Set the command type for servo.
+  servo.incomingCommandType(CommandType::POSE);
+
+  // 1.  The workflow for when command is not in planning frame.
```
```suggestion
  // 1. The workflow for when command is not in planning frame.
```

sea-bass commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp:
```diff
@@ -0,0 +1,148 @@
+  // Set tolerances for the target.
+  const double linear_tolerance{ 0.001 }, angular_tolerance{ 0.01 };
+
+  auto move_to_pose = [&](const auto& target_pose) {
```
Personal preference, but this lambda is pretty long and at this point I'd prefer a regular function.

Also, should this have a timeout just in case?

ibrahiminfinite replied:
Have reduced the size of this lambda, making it into a function in the demo would require passing around 5 arguments, which felt more complex than a slightly long lambda.

ibrahiminfinite replied:
Have added a timeout as well.

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,136 @@
+  /**
+   * \brief Computes the required change in joint angles for given cartesian change, using the robots IK solver.
+   * @param carteisan_position_delta The change in cartesian position.
+   * @return The required joing angle deltas.
```
```suggestion
   * @return The required joint angle deltas.
```

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,136 @@
+  const std::string getStatusMessage();
+
+  /**
+   * \brief Computes the required change in joint angles for given cartesian change, using the robots IK solver.
```
```suggestion
   * \brief Computes the required change in joint angles for given Cartesian change, using the robot's IK solver.
```

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,136 @@
+
+private:
+  /**
+   * \brief Set the IK solver that servo will use. If the robot does not have one, inverse jacobian will be used instead.
```
nit: but Cartesian and Jacobian are both capitalized.

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,136 @@
+
+  /**
+   * \brief Computes the required change in joint angles for given cartesian change, using the robots IK solver.
+   * @param carteisan_position_delta The change in cartesian position.
```
```suggestion
   * @param cartesian_position_delta The change in Cartesian position.
```

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/command_processor.hpp:
```diff
@@ -0,0 +1,136 @@
+   * @param carteisan_position_delta The change in cartesian position.
+   * @return The required joing angle deltas.
+   */
+  Eigen::VectorXd deltaFromIkSolver(const Eigen::VectorXd& cartesian_position_delta);
```
for consistency call this `jointDeltaFromIKSolver`?

sea-bass reviewed (commented):
Took a first pass through!

ibrahiminfinite commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -263,80 +264,91 @@ servo:
 
 ############### POSE TRACKING PARAMETERS #########################
```
These have been removed.

ibrahiminfinite commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,101 @@
+
+using namespace moveit_servo;
+
+const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_servo.joint_jog_demo");
```
All loggers have been put into anonymous namespaces.

sea-bass commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,413 @@
+  // Load the smoothing plugin
+  setSmoothingPlugin();
+
+  // Check if the tansforms to planning frame and end-effector frame exists.
```
```suggestion
  // Check if the transforms to planning frame and end effector frame exist.
```

sea-bass commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,255 @@
+
+  // Create incoming velocities with only joint 1 and joint 7 over limit by a factor of 0.1 and 0.05
+  // Scale all other joint velocities by 0.3 to keep it within limits.
+  incoming_velocities(0) += (incoming_velocities(0) * 0.1);
```
Can you do e.g. this instead?

```
incoming_velocities(0) *= 1.1;
```

ibrahiminfinite replied:
Yup, was going to change it but looks like I forgot. Thanks!

ibrahiminfinite replied:
Done

sea-bass commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,255 @@
+  incoming_velocities(5) -= (incoming_velocities(5) * 0.3);
+  incoming_velocities(6) += (incoming_velocities(6) * 0.05);
+
+  // The resulting scaling factor selected should be 0.9, but will be 0.90909094 due to float precision.
+  double user_velocity_override = servo_param_listener_->get_params().override_velocity_scaling_factor;
+  double scaling_factor =
+      moveit_servo::velocityScalingFactor(incoming_velocities, joint_bounds, user_velocity_override);
+  ASSERT_FLOAT_EQ(scaling_factor, 0.90909094);
```
You should do

```
ASSERT_NEAR(scaling_factor, 0.9, tol);
```

with `constexpr double tol = 0.01;` or something

ibrahiminfinite replied:
My bad, it is actually supposed to be 0.90909~ ( i.e 1 / 1.1) not 0.9 
Have changed the float checks to use ASSERT_NEAR.
Thanks.

sea-bass commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,255 @@
+
+  // Home state
+  Eigen::VectorXd home_state(7);
+  home_state << 0.0, -0.785, 0.0, -2.356, 0.0, 1.571, 0.785;
```
Can you get this directly from a named state from the `robot_state_` variable? No big deal either way.

ibrahiminfinite replied:
YES ! , the configuration is available as the named state "ready". 
Thanks for pointing it out.

sea-bass commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,255 @@
+  // Pose message for a carteisan delta with only a +45 degree rotation about z.
+  Eigen::VectorXd cartesian_delta(6);
+  cartesian_delta << 0.0, 0.0, 0.0, 0.0, 0.0, M_PI / 4;
+  geometry_msgs::msg::Pose recieved_pose =
```
typo: should be `received_pose`

sea-bass commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,255 @@
+  ee_pose.rotate(Eigen::AngleAxisd(M_PI / 4, Eigen::Vector3d::UnitZ()));
+  Eigen::Quaterniond ee_rotation(ee_pose.rotation());
+  double ee_pose_z = ee_rotation.z();
+  ASSERT_FLOAT_EQ(recieved_pose.orientation.z, ee_pose_z);
```
same here, use `ASSERT_NEAR` instead

sea-bass reviewed (commented):
Tests looking nice! Took another look and have a few suggestions.

I guess one of my biggest questions on the tests is... given that we took care to make a standalone C++ API, is there no way to write unit tests that do not require standing up a ROS node? Like, see here for example for a way to get a robot model: https://github.com/PickNikRobotics/pick_ik/blob/main/tests/robot_tests.cpp#L97C1-L100

Not sure how one would load an IK plugin in the similar way. Maybe something like this? https://github.com/ros-planning/moveit2/blob/main/moveit_kinematics/test/test_kinematics_plugin.cpp

ibrahiminfinite commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,412 @@
+  Eigen::Map<Eigen::VectorXd> current_joint_velocities(current_state.velocities.data(), num_joints_);
+  Eigen::Map<Eigen::VectorXd> target_joint_velocities(target_state.velocities.data(), num_joints_);
+
+  // Update filter state
+  smoother_->reset(current_state.positions);
```
It should.
If we don't update the smoother every iteration, there will be jumps when servo resumes after the robot state was changed by some other agent.
The current `servo_calcs` also does this [here](https://github.com/ros-planning/moveit/blob/69ca441bc343c51544b19dde9c5a095481a3d906/moveit_ros/moveit_servo/src/servo_calcs.cpp#L312) and [here](https://github.com/ros-planning/moveit/blob/69ca441bc343c51544b19dde9c5a095481a3d906/moveit_ros/moveit_servo/src/servo_calcs.cpp#L429)

sea-bass replied:
I see -- I think I confused `reset()` vs. `initialize()` and was thinking it was the other.

Weird naming, then.

ibrahiminfinite commented on moveit_ros/moveit_servo/tests/config/panda_simulated_config.yaml:
```diff
@@ -29,13 +30,19 @@ publish_joint_accelerations: false
 ## Other frames
-ee_frame_name: panda_hand  # The name of the end effector link, used to return the EE pose
-robot_link_command_frame:  panda_hand  # commands must be given in the frame of a robot link. Usually either the base or end effector
+ee_frame_name: panda_link8  # The name of the end effector link, used to return the EE pose
+robot_link_command_frame:  panda_link0  # commands must be given in the frame of a robot link. Usually either the base or end effector
```
One thing to note here is , the new implementation does not use the `robot_link_command_frame` parameter since the command frame is always included with the command.

ibrahiminfinite commented:
> is there no way to write unit tests that do not require standing up a ROS node?

The tests for the `utils` can be done without having to setup a ROS node.
The integration tests for jointjog, twist and pose however still need a `node` instance. (used for `transform_buffer` and `smoother` by `Servo`), and also for the `PlanningSceneMonitor`.
Since a node is required anyway, I was planning to use the same fixture for the integration tests as well, less stuff to manage.

ibrahiminfinite commented on moveit_ros/moveit_servo/tests/test_independent_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+
+
```
Done

ibrahiminfinite commented on moveit_ros/moveit_servo/tests/test_independent_utils.cpp:
```diff
@@ -0,0 +1,78 @@
+TEST(ServoUtilsUnitTests, validVector)
+{
+  Eigen::VectorXd valid_vector(7);
+  valid_vector << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0;
```
Not possible if using `VectorXd`,  but since we already know the number of joints for panda robot, I changed this to 
`vector<double, 7>` which will let us initialize directly.

AndyZe commented:
Looking at the most recent commit (24f907 Fix twist frame conversion), I have some questions.

```
// We only convert the linear velocities to planning frame, the angular velocities should be applied in the command frame itself.
...
// Transform translation to end effector frame and apply.
...
ee_pose.rotate(target_rotation);
```

^ I believe this only works if the EE frame and the command frame are the same. Are we OK with that? Can you think of any unusual use cases where that wouldn't work.

I can think of one. Before, we talked about the case where somebody is using a VR hand controller to move the robot, like [this](https://www.youtube.com/watch?v=b8K2e3Y7QtI)[ video](https://www.youtube.com/watch?v=b8K2e3Y7QtI). If we go with this approach, the user would be necessary for transforming the twist before they publish it to Servo. Otherwise, it won't be correct. I guess that's OK, it just requires extra work from the user.

I'd suggest searching the codebase for "command frame" and changing the text to "ee frame". It will be confusing to have 2 names for the same thing.

AndyZe commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,219 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, PickNik Inc.
```
```suggestion
 *  Copyright (c) 2023, PickNik Inc.
```

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/status_codes.hpp:
```diff
@@ -48,22 +50,22 @@ enum class StatusCode : int8_t
-  JOINT_BOUND = 5,
-  DECELERATE_FOR_LEAVING_SINGULARITY = 6,
-  PAUSED = 7
+  HALT_FOR_SINGULARITY = 3,
+  DECELERATE_FOR_LEAVING_SINGULARITY = 4,
+  DECELERATE_FOR_COLLISION = 5,
+  HALT_FOR_COLLISION = 6,
+  JOINT_BOUND = 7,
+  POSE_ACHIEVED = 8
```
Did you skip 2 on accident?  ;)

```suggestion
  HALT_FOR_SINGULARITY = 2,
  DECELERATE_FOR_LEAVING_SINGULARITY = 3,
  DECELERATE_FOR_COLLISION = 4,
  HALT_FOR_COLLISION = 5,
  JOINT_BOUND = 6,
  POSE_ACHIEVED = 7
```

ibrahiminfinite replied:
I think had added another status which I assigned 2, forgot to adjust the numbering after I removed it. Thanks

AndyZe commented:
Can you add a test where a twist command has the wrong frame and it gets rejected by Servo?

ibrahiminfinite commented:
> Can you add a test where a twist command has the wrong frame and it gets rejected by Servo?

Have added the rejection test for both twist and pose.

ibrahiminfinite commented:
> Looking at this class again, I wonder if it would be at all possible to change it from a class to a bunch of free functions... but maybe MoveIt makes this really hard to do this because you need to have an IK solver, robot model, robot state, etc. Just something to think about.

@sea-bass @AndyZe 
I finally found a way to do a clean  refactor of the `CommandProcessor` logic into free functions.
Things look much better now and should be easy to test as well.

A part of the motivation for the refactor was this [post](https://github.com/ros-planning/moveit2/issues/917) by @tylerjw

AndyZe commented on moveit_ros/moveit_servo/launch/ros_demo_joint_jog.launch.py:
```diff
@@ -0,0 +1,120 @@
+import os
```
Noting that this file doesn't seem to do anything yet.

I'd recommend pulling `servo_keyboard_input.cpp` out of the tutorials and putting it here instead, so users have something convenient to test with.

https://github.com/ros-planning/moveit2_tutorials/blob/main/doc/examples/realtime_servo/src/servo_keyboard_input.cpp

ibrahiminfinite replied:
> I'd recommend pulling `servo_keyboard_input.cpp` out of the tutorials and putting it here instead, so users have something convenient to test with.

That sounds like a good idea.
I already have an updated version of the `servo_keyboard_input.cpp`.

ibrahiminfinite commented:
> How did you get around that? Does it just start automatically right away?

When  the `ServoNode`  is created, the servoing thread is automatically run.
But no inputs are accepted until you set an input type.

In addition, the pause service can be used to stop accepting commands anytime.
We no longer have the concept of `start` and `stop`.

ibrahiminfinite commented:
Keyboard control demo has been moved into the demos.
After this change, the moveit2_tutorials folder for `real_time` servo no longer contains any useful code. https://github.com/ros-planning/moveit2_tutorials/tree/main/doc/examples/realtime_servo

ibrahiminfinite commented:
> Maybe some topic/namespace inconsistency?

Ah! 
That is probably because you didn't set the input type using the service
```
ros2 service call /moveit_servo/switch_command_type moveit_msgs/srv/ServoCommandType "{command_type: 2}"

```

0 for jointjog
1 for twist
2 for pose

sea-bass commented:
Aha! Should these services be called by the keyboard node itself as part of the keyboard presses, then? Especially when switching between modes?

ibrahiminfinite commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -14,16 +14,9 @@ servo:
   }
 
-  ee_frame_name: {
+  ee_frame: {
```
In that case do we also change the `planning_frame` to `planning_frame_name` ?
Feels weird otherwise since both are referring to the frame names.

sea-bass replied:
gah, I see the naming was inconsistent already in the existing YAML. In that case, let's keep it as `ee_frame` and we'll handle the migrations...

AndyZe replied:
`planning_frame` parameter is going to disappear anyway, right?

ibrahiminfinite replied:
> `planning_frame` parameter is going to disappear anyway, right?

If we are going to assume all commands are in the "IK tip frame"  (might not be same as the end effector (TCP) )
Then there is no need for either `planning_frame` or `ee_frame_name`  parameters right?

AndyZe replied:
that seems ok to me

ibrahiminfinite replied:
Keeping both parameters as `planning_frame` and `ee_frame` since we have decided not to go with IK tip only commands.

ibrahiminfinite commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -261,82 +243,42 @@ servo:
-    default_value: 0.0,
-    description: "Integral gain value for angular control"
-  }
+    frequency: {
+      type: double,
+      default_value: 50,
+      description: "The frequency for the pose tracking loop",
+      validation: {
+        gt<>: 0
+      }
+    }
```
This is something I myself had questioned.
What is a more meaningful parameter ?  Should `Servo` itself use `frequency` ?
Was there any reason for `Servo`  using `publish_period` rather than a frequency ? @AndyZe

AndyZe replied:
no, i don't care either way

ibrahiminfinite replied:
The 'frequency` and `timeout` parameters of pose tracking has been removed, since there is no longer a pose tracking thread, `Servo` responds to pose commands in the same way it handles twist and jointjog

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,114 @@
+  joint_jog << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0;
+
+  // Frquency at which the commands will be send to robot controller.
+  rclcpp::WallRate rate(1.0 / servo_params.publish_period);
```
here's the reason why it should be a WallRate:  https://github.com/ros-planning/moveit2/pull/1543

ibrahiminfinite commented on moveit_ros/moveit_servo/include/moveit_servo/utils/datatypes.hpp:
```diff
@@ -36,34 +31,96 @@
+// The joint jog command, this will be vector of length equal to the number of joints of the robot.
+typedef Eigen::VectorXd JointJog;
+
+// The twist command,  frame_id is the name of the frame in which the command is specified in.
+// frame_id must always be specified.
+struct Twist
+{
+  std::string frame_id;
+  Eigen::Vector<double, 6> velocities;
+};
+
+// The Pose command,  frame_id is the name of the frame in which the command is specified in.
+// frame_id must always be specified.
+struct Pose
+{
+  std::string frame_id;
+  Eigen::Isometry3d pose;
+};
```
yeah, that will be good.
Lets go with it. `JointJogCommand` , `TwistCommand` and `PoseCommand`.
Might help keep things clearer in cases like when a `geometry_msgs::msg::Twist` is being used without the namespace prefix.

ibrahiminfinite replied:
The above mentioned change has been applied.

sea-bass commented:
> > Aha! Should these services be called by the keyboard node itself as part of the keyboard presses, then? Especially when switching between modes?
> 
> Not sure about calling them automatically. Adding keys to call the input switching service might be good.

I think I meant that... like, there are keys to switch between endpoint and joint jogging, so those could call the respective services to change modes, right?

ibrahiminfinite commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -89,30 +89,12 @@ servo:
   enable_parameter_update: {
     type: bool,
-    default_value: false,
+    default_value: true,
     description: "If true, continuous check for parameter update is enabled \
                   Currently only the following parameters are updated:
-                  1. override_velocity_scaling_factor
-                  2. robot_link_command_frame"
+                  1. override_velocity_scaling_factor"
```
I thought it was still in discussion with the PickNik team.
Will remove it if that is okay. One less check happening continuously.

ibrahiminfinite replied:
Servo will now always check for parameter updates

ibrahiminfinite commented on moveit_ros/moveit_servo/include/moveit_servo/utils/common.hpp:
```diff
@@ -0,0 +1,160 @@
+namespace moveit_servo
+{
+
+/**
+ * \brief Checks if a given command is valid.
+ * @param command The command to be checked.
+ * @return True if the command is valid, else False.
+ */
+bool isValidCommand(const Eigen::VectorXd& command);
+
+/**
+ * \brief Checks if a given command is valid.
+ * @param command The command to be checked.
+ * @return True if the command is valid, else False.
+ */
+bool isValidCommand(const Eigen::Isometry3d& command);
+
+/**
+ * \brief Checks if a given Twist command is valid.
+ * @param command The command to be checked.
+ * @return True if the command is valid, else False.
+ */
+bool isValidCommand(const Twist& command);
+
+/**
+ * \brief Checks if a given Pose command is valid.
+ * @param command The command to be checked.
+ * @return True if the command is valid, else False.
+ */
+bool isValidCommand(const Pose& command);
```
Not sure what you mean here.
Are you pointing to the `VectorXd` and `Isometry3d` overloads ?

ibrahiminfinite replied:
Have added more info in the docstrings

ibrahiminfinite commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -89,30 +89,13 @@ servo:
     description: "If true, continuous check for parameter update is enabled \
                   Currently only the following parameters are updated:
                   1. override_velocity_scaling_factor
                   2. robot_link_command_frame"
```
Marking this and following set of comments as resolved since these change were reverted.

ibrahiminfinite commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -59,6 +52,13 @@ servo:
+
+  pose_command_in_topic: {
+    type: string,
+    default_value: "~/pose_cmd",
```
Updated to `~/pose_target_cmds`

ibrahiminfinite commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,114 @@
+  JointJog joint_jog(7);
+  joint_jog << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0;
+
+  // Frquency at which the commands will be send to robot controller.
```
Fixed

ibrahiminfinite commented on moveit_ros/moveit_servo/demos/servo_keyboard_input.cpp:
```diff
@@ -0,0 +1,313 @@
+const std::string JOINT_TOPIC = "/moveit_servo/delta_joint_cmds";
+const size_t ROS_QUEUE_SIZE = 10;
+const std::string EEF_FRAME_ID = "panda_hand";
+const std::string BASE_FRAME_ID = "panda_link0";
```
The `#define` s have been removed.
Two new keys `j`  for  jointjog and `t`  for twist have been added to call the service to switch the input type.

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp:
```diff
@@ -0,0 +1,143 @@
+ *      Description : Example of controlling a robot through pose commands via the C++ API.
+ */
+
+#include <chrono>
+#include <atomic>
+#include <mutex>
+#include <rclcpp/rclcpp.hpp>
+#include <moveit_servo/servo.hpp>
+#include <moveit_servo/utils/common.hpp>
```
Nitpick: can you go through and sort the includes alphabetically? Makes for better readability. Your IDE should have a hotkey to do this quickly.

ibrahiminfinite replied:
Done !

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp:
```diff
@@ -0,0 +1,143 @@
+  rclcpp::WallRate command_rate(50);
+  RCLCPP_INFO_STREAM(LOGGER, servo.getStatusMessage());
+
+  while (!stop_tracking)
```
```suggestion
  while (!stop_tracking && rclcpp::ok())
```

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp:
```diff
@@ -0,0 +1,143 @@
+  auto pose_tracker = [&]() {
+    KinematicState joint_state;
+    rclcpp::WallRate tracking_rate(1 / servo_params.publish_period);
+    while (!stop_tracking)
```
To ensure Ctrl+C still functions. Please check if this change is needed anywhere else.

```suggestion
    while (!stop_tracking && rclcpp::ok())
```

ibrahiminfinite replied:
Done

AndyZe commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp:
```diff
@@ -0,0 +1,143 @@
+      std::lock_guard<std::mutex> pguard(pose_guard);
+      target_pose.pose = servo.getEndEffectorPose();
+      stop_tracking = target_pose.pose.isApprox(terminal_pose, servo_params.pose_tracking.linear_tolerance);
+      target_pose.pose.translate(Eigen::Vector3d(0.0, 0.0, -delta));
```
Is this comment correct? I think it would be nice to highlight dynamic pose tracking.

```suggestion
      // Dynamically update the target pose
      target_pose.pose.translate(Eigen::Vector3d(0.0, 0.0, -delta));
```

ibrahiminfinite replied:
Yeah, makes it more clear. 👍

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/collision_monitor.hpp:
```diff
@@ -1,12 +1,7 @@
  * BSD 3-Clause License
  *
- * Copyright (c) 2019, Los Alamos National Security, LLC
+ * Copyright (c) 2023, PickNik Inc.
```
Changing license holders is not cool.

```suggestion
 * Copyright (c) 2019, Los Alamos National Security, LLC
```

ibrahiminfinite replied:
oh, sorry about that !!

AndyZe reviewed (commented):
This looks close to mergeable! Thanks for a lot of great work.

- [ ] In Cartesian mode of the ROS api demo -- motion of the robot seems very slow. Can you make it about 4X faster?

- [x] ros2 launch moveit_servo demo_twist.launch.py -->

[demo_twist-5] [WARN] [1689438088.851520932] [moveit_servo.command_processor]: Command frame is: panda_link8 expected: panda_link0

ibrahiminfinite commented on moveit_ros/moveit_servo/demos/servo_keyboard_input.cpp:
```diff
@@ -0,0 +1,356 @@
+  puts("Reading from keyboard");
+  puts("---------------------------");
+  puts("Use arrow keys and the '.' and ';' keys to Cartesian jog");
+  puts("Use 'w' to Cartesian jog in the world frame, and 'e' for the End-Effector frame");
```
Yeah, those can be removed. 
Thanks.

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:
```diff
@@ -0,0 +1,194 @@
+  void setCollisionChecking(const bool check_collision);
+
+  /**
+   * \brief Converts the given pose to planning frame.
```
```suggestion
   * \brief Converts the given pose to planning frame. Assumes the command is in a stationary reference frame (issue #2150)
```

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+geometry_msgs::msg::Pose poseFromCartesianDelta(const Eigen::VectorXd& delta_x,
+                                                const Eigen::Isometry3d& base_to_tip_frame_transform)
+{
+  // get a transformation matrix with the desired position change &
+  // get a transformation matrix with desired orientation change
```
Maybe make these comments less redundant, e.g.

```suggestion
  // Get separate transformation matrices with the desired position and orientation change.
```

sea-bass replied:
that was another suggestion I was considering. I'm fine with that too

sea-bass commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,238 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, PickNik Inc.
```
Doesn't seem like this was resolved?

sea-bass reviewed (commented):
I have great news. I built your forks of `moveit2` and `moveit_msgs` against MoveIt Studio, so here is this PR doing Cartesian and joint jogging perfectly.

https://github.com/ros-planning/moveit2/assets/4603398/593f1ee6-8b9c-4849-838c-86c45d9d0cf2

**NOTE**: I did have to change the servo `ee_frame_name` to `ee_frame` for this all to work, so I have a question about whether we want to help the users transition through e.g. a deprecation strategy.

Besides that, just some minor stuff, and I'm happy to approve after.

sea-bass commented:
... wait a second.

I just searched the new source code and found nothing tied to the `command_in_type` parameter and the choice between `unitless` and `speed_units`. Was this feature removed? And if so, we need to bring it back.

ibrahiminfinite commented on moveit_ros/moveit_servo/include/moveit_servo/servo_node.hpp:
```diff
@@ -0,0 +1,123 @@
+
+  /**
+   * \brief The service to set the command type for Servo.
+   * Supported commands are
+   * 0 - JointJog
+   * 1 - Twist
+   * 2 - Pose
```
That sounds good 👍

ibrahiminfinite commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,192 @@
+    if (expectedType == CommandType::JOINT_JOG && new_joint_jog_msg_)
+    {
+      new_joint_jog_msg_ = false;
+      // JointJog is an alias for VectorXd, so we can directly make a map and pass it.
```
Ohh, got leftover during the update of typedef name from JointJog -> JointjogCommand

ibrahiminfinite commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,457 @@
+  return params_valid;
+}
+
+const Eigen::Isometry3d Servo::getEndEffectorPose()
```
The function was kept intentionally, (mostly a convenience function)  so that users can write control logic like in the pose demo if they need to.

ibrahiminfinite commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -14,16 +14,9 @@ servo:
   }
 
-  ee_frame_name: {
+  ee_frame: {
```
> should we just move on this from and deal with issues that may come in?
I am in favour of this.

ibrahiminfinite commented on moveit_ros/moveit_servo/launch/demo_joint_jog.launch.py:
```diff
@@ -84,19 +77,6 @@ def generate_launch_description():
         composable_node_descriptions=[
-            # Example of launching Servo as a node component
-            # Launching as a node component makes ROS 2 intraprocess communication more efficient.
-            launch_ros.descriptions.ComposableNode(
```
Yeah, will update it run as node component :+1:

ibrahiminfinite replied:
`ServoNode` can now be launched as a component.
The C++ API demos are still launched as standalone, not sure if making the demos composable is necessary.

ibrahiminfinite commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,457 @@
+  }
+  else
+  {
+    RCLCPP_INFO(LOGGER, "IK solver available for robot, will use it.");
```
Unfortunately it does not look like that information is not available from the `KinematicsBase`, maybe this is something worth adding ?

sea-bass commented on moveit_ros/moveit_servo/src/collision_monitor.cpp:
```diff
@@ -0,0 +1,154 @@
+namespace moveit_servo
+{
+
+CollisionMonitor::CollisionMonitor(const planning_scene_monitor::PlanningSceneMonitorPtr& planning_scene_monitor,
```
We got an interesting request about this...

Is there a way we can have this class have a standalone function that can be called in other code that accepts a robot state and/or twist and returns to you the collision scaling?

So basically create a method that is called in the loop of `checkCollisions()` that can also be called by itself? Also lets you unit test easily.

AndyZe replied:
IMO this is the type of feature that can be worked on later. Can't/shouldn't put too much on @ibrahiminfinite's TODO list at once.

sea-bass replied:
I completely agree!

ibrahiminfinite replied:
Yeah, I'd be happy to get these done once this main PR goes in :+1:

sea-bass commented on moveit_ros/moveit_servo/config/panda_simulated_config.yaml:
```diff
@@ -2,22 +2,13 @@
 publish_period: 0.034  # 1/Nominal publish rate [seconds]
-low_latency_mode: false  # Set this to true to publish as soon as an incoming Twist command is received (publish_period is ignored)
 
+command_in_type: "speed_units"
```
Maybe bring back a comment above this parameter since this example config is meant to help people understand their options

sea-bass commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -162,15 +161,6 @@ servo:
                   this should be set to false"
   }
 
-  incoming_command_timeout: {
-    type: double,
-    default_value: 0.1,
-    description: "Stop servoing if X seconds elapse without a new command. \
-                  If 0, republish commands forever even if the robot is stationary. \
-                  Otherwise, specify num. to publish. Important because ROS may drop \
-                  some messages and we need the robot to halt reliably."
-  }
-
```
Is this timeout gone now? Does something else replace it?

sea-bass commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -221,12 +211,6 @@ servo:
     }
   }
 
-  low_latency_mode: {
-    type: bool,
-    default_value: false,
-    description: "If true , low latency mode is enabled."
-  }
```
We're currently using this parameter on MoveIt Studio for Gazebo, but it's kind of a workaround and I'm OK removing this parameter and seeing what happens / looking for better solutions.

ibrahiminfinite replied:
We can add this later if necessary :+1:

sea-bass commented on moveit_ros/moveit_servo/config/test_config_panda.yaml:
```diff
@@ -2,20 +2,17 @@
-publish_period: 0.034  # 1/Nominal publish rate [seconds]
+publish_period: 0.02  # 1/Nominal publish rate [seconds]
+
+command_in_type: "unitless"
+scale:
+  linear: 1.0
+  rotational: 1.0
+  joint: 1.0
```
Bring back comments for all these parameters, as in the old version

sea-bass commented on moveit_ros/moveit_servo/config/test_config_panda.yaml:
```diff
@@ -27,18 +24,21 @@ publish_joint_velocities: true
-ee_frame_name: panda_hand  # The name of the end effector link, used to return the EE pose
-robot_link_command_frame:  panda_hand  # commands must be given in the frame of a robot link. Usually either the base or end effector
-
-## Stopping behaviour
-incoming_command_timeout:  0.1  # Stop servoing if X seconds elapse without a new command
```
Again, is this functionality gone?

sea-bass commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_joint_jog.cpp:
```diff
@@ -0,0 +1,114 @@
+  JointJogCommand joint_jog(7);
+  joint_jog << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0;
+
+  // Frequency at which the commands will be send to robot controller.
```
```suggestion
  // Frequency at which commands will be sent to the robot controller.
```

sea-bass commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp:
```diff
@@ -0,0 +1,153 @@
+  std::thread tracker_thread(pose_tracker);
+  tracker_thread.detach();
+
+  const double delta = 0.002;
```
What is `delta`? Consider adding a comment and a more descriptive name explaining what this does -- which is dynamically updating the target pose so it's moving up in the +Z direction at a particular rate.

In fact, I might say bring up the entire `Eigen::Vector3d` defining the translation up in front of the loop so people can more easily change this value. Also, consider making it an `Isometry3d` instantiated using both translation and euler angles for the same reason. So people can change the demo to have the target pose rotate as well.

As an example, see:

```
Eigen::Isometry3d const goal_frame =
            Eigen::Translation3d(std::sin(M_PI_4), 3.0 * std::sin(M_PI_4), 0.0) *
            Eigen::AngleAxisd(0.75 * M_PI, Eigen::Vector3d::UnitZ());
```

ibrahiminfinite replied:
Done.

sea-bass commented on moveit_ros/moveit_servo/demos/cpp_interface/demo_twist.cpp:
```diff
@@ -0,0 +1,146 @@
+  tf2::doTransform(transformed_twist, transformed_twist, command_to_planning_frame);
+  target_twist = TwistCommand{ servo_params.planning_frame, transformed_twist };
+
+  // Frequency at which the commands will be send to robot controller.
```
```suggestion
  // Frequency at which commands will be sent to the robot controller.
```

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/utils/common.hpp:
```diff
@@ -0,0 +1,158 @@
+                                       const moveit::core::JointBoundsVector& joint_bounds, double scaling_override);
+
+/**
+ * \brief Finds the joints that are exceeding allowable joint limits.
+ * @param positions The joints positions.
+ * @param velocities The current commanded velocities.
+ * @param joint_bounds The bounding information for the robot joints.
+ * @param margin Additional buffer on the actual joint limits.
+ * @return The joints that are violating the specified position limits.
```
```suggestion
 * \brief Finds the joints that are exceeding allowable limits.
 * @param positions The joint positions.
 * @param velocities The current commanded velocities.
 * @param joint_bounds The bounding information for the robot joints.
 * @param margin Additional buffer on the actual joint limits.
 * @return The joints that are violating the specified limits.
```

it's not just position limits that can be exceeded, right?

ibrahiminfinite replied:
`jointsToHalt` only checks if the commanded velocity will make the joint go past position limit.

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/utils/common.hpp:
```diff
@@ -0,0 +1,158 @@
+ * @param eigen_tf The isometry to be converted to TransformStamped.
+ * @param parent_frame The target frame.
+ * @param child_frame The current frame.
+ * @return The isometery as TransformStamped message.
```
```suggestion
 * @return The isometry as a TransformStamped message.
 ```

sea-bass commented on moveit_ros/moveit_servo/src/collision_monitor.cpp:
```diff
@@ -0,0 +1,154 @@
+
+  while (rclcpp::ok() && !stop_requested_)
+  {
+    if (servo_params_.check_collisions)
```
There's no `else` to this which simply sets the scaling to 1.0 -- should there be?

ibrahiminfinite replied:
Can be fixed by setting `collision_velocity_scale_` to 1.0 at the beginning of every loop :+1:

sea-bass commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,404 @@
+    // Compute velocities based on smoothed joint positions
+    target_joint_velocities = (target_joint_positions - current_joint_positions) / servo_params_.publish_period;
+
+    // TODO : print warning if scaling applied for joint limit.
```
Is this TODO for this PR? Seems like a quick one, so probably?

ibrahiminfinite replied:
Done

sea-bass commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,404 @@
+    target_joint_positions = current_joint_positions + (target_joint_velocities * servo_params_.publish_period);
+
+    // Check if any joints are going past joint position limits
+    std::vector<int> joints_to_halt =
```
const?

sea-bass commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,404 @@
+
+  // Get necessary information about joints
+  const std::vector<std::string> joint_names = joint_model_group->getActiveJointModelNames();
+  moveit::core::JointBoundsVector joint_bounds = joint_model_group->getActiveJointModelsBounds();
```
Should this be const?

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,220 @@
+namespace
+{
+const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_servo.servo_node");
+// This value is used when configuring the main loop to use SCHED_FIFO scheduling
+// We use a slightly lower priority than the ros2_control default in order to reduce jitter
+// Reference: https://man7.org/linux/man-pages/man2/sched_setparam.2.html
+int const THREAD_PRIORITY = 40;
```
This addition seems a bit hacky... could it at least be configurable?

ibrahiminfinite replied:
Updated this to be configurable parameter, will be useful to set priority without recompiling.

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,220 @@
+
+    if (expectedType == CommandType::JOINT_JOG && new_joint_jog_msg_)
+    {
+      // Mark latest jointjog command as processed and also reject any twist and pose messages that had arrived simultaneously.
```
For maintanability, I would suggest not listing everything out and saying something like

```suggestion
      // Mark latest joint jog command as processed and also reject any other command types that had arrived simultaneously.
```

and same for the other comments like it here

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,220 @@
+
+    servo_params_ = servo_->getParams();
+    publish_command = false;
+    CommandType expectedType = servo_->expectedCommandType();
```
const?

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,220 @@
+      // Mark latest jointjog command as processed and also reject any twist and pose messages that had arrived simultaneously.
+      new_joint_jog_msg_ = new_twist_msg_ = new_pose_msg_ = false;
+      // JointJogCommand is an alias for VectorXd, so we can directly make a map and pass it.
+      Eigen::Map<JointJogCommand> command(latest_joint_jog_.velocities.data(), latest_joint_jog_.velocities.size());
```
const?

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,220 @@
+    {
+      // Mark latest twist command as processed and also reject any jointjog and pose messages that had arrived simultaneously.
+      new_joint_jog_msg_ = new_twist_msg_ = new_pose_msg_ = false;
+      Eigen::Vector<double, 6> velocities{ latest_twist_.twist.linear.x,  latest_twist_.twist.linear.y,
+                                           latest_twist_.twist.linear.z,  latest_twist_.twist.angular.x,
+                                           latest_twist_.twist.angular.y, latest_twist_.twist.angular.z };
+      TwistCommand command{ latest_twist_.header.frame_id, velocities };
```
const for both of these?

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,220 @@
+    {
+      // Mark latest pose command as processed and also reject any jointjog and twist messages that had arrived simultaneously.
+      new_joint_jog_msg_ = new_twist_msg_ = new_pose_msg_ = false;
+      PoseCommand command = poseFromPoseStamped(latest_pose_);
```
const?

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,133 +31,220 @@
+      trajectory_publisher_->publish(composeTrajectoryMessage(servo_params_, next_joint_states));
+      publish_command = false;
+    }
+    auto status_msg = std::make_unique<std_msgs::msg::Int8>();
+    status_msg->data = static_cast<int8_t>(servo_->getStatus());
+    status_publisher_->publish(std::move(status_msg));
```
Is there a need to make this a unique pointer vs. just making the Int8 message and not having to dereference/move it to publish?

ibrahiminfinite replied:
That was copy-pasta from the old implementation.
Now that I look at the [documentation](https://docs.ros2.org/latest/api/rclcpp/classrclcpp_1_1Publisher.html), it seems creating the status once on the stack is better, since it takes a reference.

```
virtual void rclcpp::Publisher< MessageT, AllocatorT >::publish(const MessageT &  msg)
```

sea-bass commented on moveit_ros/moveit_servo/src/utils/command.cpp:
```diff
@@ -0,0 +1,239 @@
+{
+  // Find the target joint position based on the commanded joint velocity
+  StatusCode status = StatusCode::NO_WARNING;
+  int num_joints = robot_state->getJointModelGroup(servo_params.move_group_name)->getJointModelNames().size();
```
const

sea-bass commented on moveit_ros/moveit_servo/src/utils/command.cpp:
```diff
@@ -0,0 +1,239 @@
+                                    servo::Params& servo_params)
+{
+  StatusCode status = StatusCode::NO_WARNING;
+  int num_joints = robot_state->getJointModelGroup(servo_params.move_group_name)->getJointModelNames().size();
```
const

sea-bass commented on moveit_ros/moveit_servo/src/utils/command.cpp:
```diff
@@ -0,0 +1,239 @@
+    // Compute linear and angular change needed.
+    const Eigen::Isometry3d ee_pose{ robot_state->getGlobalLinkTransform(servo_params.ee_frame) };
+    cartesian_position_delta.head<3>() = command.pose.translation() - ee_pose.translation();
+    Eigen::Quaterniond q_current(ee_pose.rotation()), q_target(command.pose.rotation());
+    Eigen::Quaterniond q_error = q_target * q_current.inverse();
+    Eigen::AngleAxisd angle_axis_error(q_error);
```
const for all 3?

sea-bass commented on moveit_ros/moveit_servo/src/utils/command.cpp:
```diff
@@ -0,0 +1,239 @@
+        robot_state->getGlobalLinkTransform(ik_solver->getBaseFrame()).inverse() *
+        robot_state->getGlobalLinkTransform(ik_solver->getTipFrame());
+
+    geometry_msgs::msg::Pose next_pose = poseFromCartesianDelta(cartesian_position_delta, base_to_tip_frame_transform);
```
const?

sea-bass commented on moveit_ros/moveit_servo/src/utils/command.cpp:
```diff
@@ -0,0 +1,239 @@
+  else
+  {
+    // Robot does not have an IK solver, use inverse Jacobian to compute IK.
+    Eigen::MatrixXd jacobian = robot_state->getJacobian(joint_model_group);
+    Eigen::JacobiSVD<Eigen::MatrixXd> svd =
+        Eigen::JacobiSVD<Eigen::MatrixXd>(jacobian, Eigen::ComputeThinU | Eigen::ComputeThinV);
+    Eigen::MatrixXd matrix_s = svd.singularValues().asDiagonal();
+    Eigen::MatrixXd pseudo_inverse = svd.matrixV() * matrix_s.inverse() * svd.matrixU().transpose();
```
all const?

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+
+bool isValidCommand(const Eigen::VectorXd& command)
+{
+  // returns true only if there are no nan values.
+  return (command.array() == command.array()).all();
```
can you instead use `hasNaN()` from Eigen itself?
https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#ab13d158c900560d3e1b25d85d2d33dd6

same for the other `isValidCommand` versions below

ibrahiminfinite replied:
Oh, didn't notice Eigen had these.
The `allFinite()` feels more intuitive to use here.

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+  Eigen::Matrix3d identity, rotation;
+  identity.setIdentity();
+  rotation = command.linear();
+  // checks rotation, will fail if there is nan
+  is_valid_rotation = identity.isApprox(rotation.inverse() * rotation);
+  // Command is not vald if there is Nan
+  const bool not_nan = isValidCommand(command.translation());
+  return is_valid_rotation && not_nan;
```
assuming you can't just use `hasNaN()` to do everything for you in this function, at least change `not_nan` to `is_valid_translation` to be consistent. But definitely try the first.

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+  const double leaving_singularity_threshold_multiplier = servo_params.leaving_singularity_threshold_multiplier;
+
+  // Get size of total controllable dimensions.
+  size_t dims = target_delta_x.size();
```
const?

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+  // Get size of total controllable dimensions.
+  size_t dims = target_delta_x.size();
+
+  // Get the current Jacobian and compute SVD
+  Eigen::JacobiSVD<Eigen::MatrixXd> current_svd = Eigen::JacobiSVD<Eigen::MatrixXd>(
+      robot_state->getJacobian(joint_model_group), Eigen::ComputeThinU | Eigen::ComputeThinV);
+  Eigen::MatrixXd matrix_s = current_svd.singularValues().asDiagonal();
+
+  // Compute pseudo inverse
+  Eigen::MatrixXd pseudo_inverse = current_svd.matrixV() * matrix_s.inverse() * current_svd.matrixU().transpose();
```
const?

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+  // This vector represents the least responsive dimension. By convention this is the last column of the matrix U.
+  // The sign of the singular vector from result of SVD is not reliable, so we need to do extra checking to make sure of
+  // the sign. See R. Bro, "Resolving the Sign Ambiguity in the Singular Value Decomposition".
+  Eigen::VectorXd vector_towards_singularity = current_svd.matrixU().col(dims - 1);
```
const?

ibrahiminfinite replied:
This cannot be const as it could be updated with a sign flip later.

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+  const double current_condition_number = current_svd.singularValues()(0) / current_svd.singularValues()(dims - 1);
+
+  // Take a small step in the direction of vector_towards_singularity
+  double scale = 100;
```
What is this magic number? Shouldn't this be configurable in some way, or at least put in a more visible place, kind of like that SCALING_OVERRIDE_THRESHOLD value?

ibrahiminfinite replied:
Made this into a parameter with default value as well

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+  // By convention these are the first and last element of the diagonal.
+  const double current_condition_number = current_svd.singularValues()(0) / current_svd.singularValues()(dims - 1);
+
+  // Take a small step in the direction of vector_towards_singularity
+  double scale = 100;
+  Eigen::VectorXd delta_x = vector_towards_singularity / scale;
```
const

sea-bass commented on moveit_ros/moveit_servo/src/utils/common.cpp:
```diff
@@ -0,0 +1,333 @@
+
+  // Compute the Jacobian SVD for the new robot state.
+  robot_state->setJointGroupPositions(joint_model_group, next_joint_angles);
+  Eigen::JacobiSVD<Eigen::MatrixXd> next_svd = Eigen::JacobiSVD<Eigen::MatrixXd>(
+      robot_state->getJacobian(joint_model_group), Eigen::ComputeThinU | Eigen::ComputeThinV);
```
const?

sea-bass commented on moveit_ros/moveit_servo/tests/test_utils.cpp:
```diff
@@ -0,0 +1,238 @@
+  ASSERT_EQ(scaling_result.second, moveit_servo::StatusCode::DECELERATE_FOR_LEAVING_SINGULARITY);
+}
+
+TEST(ServoUtilsUnitTests, JointLimitVeclocityScaling)
```
small typo

```suggestion
TEST(ServoUtilsUnitTests, JointLimitVelocityScaling)
```

sea-bass reviewed (changes requested):
Gave the code a thorough read through and found 4 things that I would consider necessary to resolve before getting this in. Especially 1 and 2.

1. The commands timeout parameter is gone, but I feel like that's a very useful one. Has it been replaced with something else I missed, or does it need to be brought back?
2. I didn't see anything in the collision checking logic that resets the scaling to 1 if collision checking is disabled but the loop is still going. Is this being handled in some other way, or is it a bug?
3. Lots of variables being created that could be `const`, but aren't. Give it a look.
4. Eigen has a `hasNaN()` method for its datatypes that you can probably use instead of doing things manually.

sea-bass commented:
Also the CI build failures are because a new version of `moveit_msgs` just landed on main, and the error seems related to that.

Unsure if this branch is out of date, or it's something in CI itself.

ibrahiminfinite commented:
> * The commands timeout parameter is gone, but I feel like that's a very useful one. Has it been replaced with something else I missed, or does it need to be brought back?

Ah, it did not matter in the C++ API, but is useful in the ROS API.
Will add it back :+1: 

> * I didn't see anything in the collision checking logic that resets the scaling to 1 if collision checking is disabled but the loop is still going. Is this being handled in some other way, or is it a bug?

It is a bug, must have accidentally got deleted during the last fix.
And this just reminded me that the test for collision monitor has not been added.

sea-bass commented on moveit_ros/moveit_servo/config/servo_parameters.yaml:
```diff
@@ -87,13 +89,23 @@ servo:
+    }
+  }
+
+  incoming_command_timeout: {
+    type: double,
+    default_value: 0.1,
+    description: "Commands will be discarded if it is older than the timeout.\
+                   Important because ROS may drop some messages."
```
You could keep the same comment as before? Seems the old comment had more details.

ibrahiminfinite replied:
Those additional details do not apply to the new version.
It was about "filtered halt" since the old implementation would always publish a message every iteration.

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -53,111 +76,231 @@ ServoNode::ServoNode(const rclcpp::NodeOptions& options)
+void ServoNode::switchCommandType(const std::shared_ptr<moveit_msgs::srv::ServoCommandType::Request> request,
+                                  const std::shared_ptr<moveit_msgs::srv::ServoCommandType::Response> response)
+{
+  const bool is_valid = (request->command_type >= 0) && (request->command_type <= 2);
```
Use the enum names instead of hard-coded numbers here.

ibrahiminfinite replied:
Done

sea-bass reviewed (commented):
I'm gonna build this branch against MoveIt Studio and get back with an approval unless I see something functionally not working. Leaving what little I could find until then... so close!

sea-bass commented:
@ibrahiminfinite I think the issue is that you need to rebase your `moveit_msgs` fork as well on the latest `ros2` branch!

sea-bass reviewed (changes requested):
Tested against MoveIt Studio.

The `/start_servo` and `/stop_servo` services are not actually started in the new node. We need those. Probably worth adding integration tests for that as well, if it makes sense.

ibrahiminfinite commented:
> Tested against MoveIt Studio (after rebasing your branches for this + moveit_msgs)
> 
> The `/start_servo` and `/stop_servo` services are not actually started in the new node. We need those. Probably worth adding integration tests for that as well, if it makes sense.
> 
> The keyboard teleop node has clients to these, so you should be able to also test interactively to verify that they work.

Are those services necessary?  They were intentionally removed and replaced by the single `/pause_servo` service.
There is also an issue linked to this PR for depreciation of those services.
I could add them in with a depreciation message if needed, but they would essentially be doing the same thing as `/pause_servo` service spilit into two.

> @ibrahiminfinite I think the CI build issue is that you need to rebase your `moveit_msgs` fork as well on the latest `ros2` branch!

I had rebased it on `ros2` branch earlier today.

sea-bass commented:
I see... if the goal is to use `/pause_servo`, and it allows pausing and unpausing in an equivalent way, that's fine. Could you share with me what needs to be done to update that?

(and generally, would be good to jot down all the incompatibilities so we can share them post release)

ibrahiminfinite commented:
> The servo loop runs automatically on Servo instance creation.
> Servo can be paused/ unpaused using the bool type service `moveit_servo/pause_servo`.
> There is no notion of start/stop in the current implementation.

I had put it in the PR description itself  , I guess there was too much information in there.
Yeah, it's just a Boolean service.
Servo is unpaused by default when in starts.
To pause , you have to call the service with value `True`

sea-bass commented:
You'll need to update the `teleop.py` client to call the `/pause_servo` service instead of start/stop on this end, BTW.

sea-bass commented:
Also, I tested again with the changes and the service shows up as

`/node_name/moveit_servo/pause_servo`

and same with `/node_name/moveit_servo/switch_command_type`

Do we have a need for adding tihs `moveit_servo` in the middle? I can always change it, but it seems this is introducing even more incompabilities than necessary.

sea-bass commented:
One more issue. You don't have a PR Up for `moveit_msgs` fork yet, right?

The `ServoCommandType` definition looks like this:

```
int8 command_type
---
int8 expected_type
```

It should also define the enumeration literals so they can be used. See for example https://github.com/ros-planning/moveit_msgs/blob/ros2/msg/MoveItErrorCodes.msg

sea-bass commented:
Once I was able to get around the issues mentioned above, I found one more bug.

It seems like servo is always publishing trajectory commands even if we set the `command_out_type` to be `std_msgs/Float64MultiArray` -- I think this switch was not ported over.

sea-bass commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,407 @@
+CommandType Servo::expectedCommandType()
 {
-  return servo_calcs_.getCommandFrameTransform(transform);
+  return expected_command_type_;
 }
 
-bool Servo::getEEFrameTransform(Eigen::Isometry3d& transform)
+void Servo::expectedCommandType(const CommandType& command_type)
 {
-  return servo_calcs_.getEEFrameTransform(transform);
+  expected_command_type_ = command_type;
 }
```
It's confusing that these both have the same name... maybe call them get/set?

ibrahiminfinite replied:
Done

ibrahiminfinite commented:
@sea-bass 
Changes from last commit:

- Updated the `ServoCommandType` service to have a `bool` response called `success` which will be `true` if command type was successfully set.
- Added a new `ServoStatus` message that contains both a status `code` and `message` explaining the meaning.
- Added enumerated items to both `ServoCommandType` and `ServoStatus` for user clarity.
- Fixed node namespace issue.
- Added support for `Float64MultiArray`


> One more issue. You don't have a PR Up for `moveit_msgs` fork yet, right?

The PR is already there https://github.com/ros-planning/moveit_msgs/pull/161

sea-bass commented on moveit_ros/moveit_servo/launch/demo_ros_api.launch.py:
```diff
@@ -85,16 +85,18 @@ def generate_launch_description():
         composable_node_descriptions=[
-            # Example of launching Servo as a node component
-            # Launching as a node component makes ROS 2 intraprocess communication more efficient.
+            # # Example of launching Servo as a node component
+            # # Launching as a node component makes ROS 2 intraprocess communication more efficient.
```
Double comments

sea-bass commented on moveit_ros/moveit_servo/launch/demo_ros_api.launch.py:
```diff
@@ -126,7 +129,7 @@ def generate_launch_description():
         ],
         output="screen",
-        condition=IfCondition(launch_as_standalone_node),
+        # condition=IfCondition(launch_as_standalone_node),
```
Re-add? Might have been testing with it

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -110,23 +110,25 @@ ServoNode::ServoNode(const rclcpp::NodeOptions& options)
-  // Create publisher for joint trajectory message
-  trajectory_publisher_ = node_->create_publisher<trajectory_msgs::msg::JointTrajectory>(
-      servo_params_.command_out_topic, rclcpp::SystemDefaultsQoS());
+  if (servo_params_.command_out_type == "trajectory_msgs/JointTrajectory")
+    trajectory_publisher_ = node_->create_publisher<trajectory_msgs::msg::JointTrajectory>(
+        servo_params_.command_out_topic, rclcpp::SystemDefaultsQoS());
+  else if (servo_params_.command_out_type == "std_msgs/Float64MultiArray")
+    multi_array_publisher_ = node_->create_publisher<std_msgs::msg::Float64MultiArray>(servo_params_.command_out_topic,
```
We should have an "else" that logs an error with invalid type -- unless this is validated elsewhere?

ibrahiminfinite replied:
It is validated by generate_parameter_library

sea-bass commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -290,10 +293,14 @@ void ServoNode::servoLoop()
     if (next_joint_state && (servo_->getStatus() != StatusCode::INVALID))
     {
-      trajectory_publisher_->publish(composeTrajectoryMessage(servo_->getParams(), next_joint_state.value()));
+      if (servo_params_.command_out_type == "trajectory_msgs/JointTrajectory")
+        trajectory_publisher_->publish(composeTrajectoryMessage(servo_->getParams(), next_joint_state.value()));
+      else if (servo_params_.command_out_type == "std_msgs/Float64MultiArray")
+        multi_array_publisher_->publish(composeMultiArrayMessage(servo_->getParams(), next_joint_state.value()));
```
This code addition is duplicated from above -- worth making it into a function?

ibrahiminfinite replied:
The one above is for publisher creation during `ServoNode` initialisation.
This is for publishing.

sea-bass replied:
wow, my bad!

I was confused because of the if/else. I think the publishers should both be created no matter what -- it doesn't hurt to have that extra publisher sitting around doing nothing, and would allow switching the command type at runtime (though not sure why someone would do that lol)

ibrahiminfinite replied:
In that case, we would have to add another `command_out_topic` parameter for the `Float64MultiArray`.
Otherwise two different type of publishers point to the same topic.
The old `servo_calcs` also use the [same](https://github.com/ros-planning/moveit2/blob/830ceda79b8f67c2224eb9ea8eb37f4aa0fe2596/moveit_ros/moveit_servo/src/servo_calcs.cpp#L109) condition based publisher creation

sea-bass replied:
Ah, I see, it's only one topic. Ignore me then, sorry!

sea-bass reviewed (commented):
Thanks for turning around these changes!

I also think there are some test failures happening in CI.

AndyZe commented on moveit_ros/moveit_servo/include/moveit_servo/utils/datatypes.hpp:
```diff
@@ -36,34 +31,99 @@
 
 namespace moveit_servo
 {
+// Datatypes used by servo
+
```
already covered by the description above
```suggestion
```

AndyZe commented on moveit_ros/moveit_servo/src/servo_node.cpp:
```diff
@@ -31,20 +31,43 @@
- *      Project   : moveit_servo
- *      Created   : 12/31/2018
- *      Author    : Andy Zelenak
+/*      Title       : servo_node.cpp
+ *      Project     : moveit_servo
+ *      Created     : 01/07/2023
+ *      Author      : V Mohammed Ibrahim
```
```suggestion
/*      Title       : servo_node.cpp
 *      Project     : moveit_servo
 *      Created     : 12/31/2018
 *      Author      : Andy Zelenak, V Mohammed Ibrahim
```

AndyZe commented on moveit_ros/moveit_servo/tests/servo_cpp_fixture.hpp:
```diff
@@ -32,20 +32,42 @@
+   Desc      : Resources used by servo c++ integration tests
+   Title     : servo_cpp_fixture.hpp
+   Project   : moveit_servo
+   Created   : 07/07/2023
```
```suggestion
   Created   : 07/13/2020
```

AndyZe commented on moveit_ros/moveit_servo/tests/servo_ros_fixture.hpp:
```diff
@@ -0,0 +1,134 @@
+  std::atomic<int> status_count_;
+  std::atomic<moveit_servo::StatusCode> status_;
+
+  std::mutex joint_state_mutex;
```
```suggestion
  std::mutex joint_state_mutex_;
```

AndyZe commented on moveit_ros/moveit_servo/tests/servo_ros_fixture.hpp:
```diff
@@ -0,0 +1,134 @@
+      if (!rclcpp::ok())
+      {
+        RCLCPP_ERROR(logger, "Interrupted while waiting for the service. Exiting.");
+      }
```
This appears to not actually exit

AndyZe replied:
I think you could just add `std::exit(EXIT_FAILURE);`. That would cause whatever test to fail

ibrahiminfinite replied:
Thanks for the catch!
There was originally a `std::exit()` there, but got deleted sometime during the changes.

AndyZe commented on moveit_ros/moveit_servo/tests/test_ros_integration.cpp:
```diff
@@ -0,0 +1,236 @@
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <geometry_msgs/msg/twist_stamped.hpp>
+
+/**
+ * Order of joints in joint state message
+ *
+ * - panda_finger_joint1
+ * - panda_joint1
+ * - panda_joint2
+ * - panda_finger_joint2
+ * - panda_joint3
+ * - panda_joint4
+ * - panda_joint5
+ * - panda_joint6
+ * - panda_joint7 --> index 8
+ */
+
```
I think they could figure this out from L96

```suggestion
```

ibrahiminfinite replied:
This was added in for L116, since `panda_joint7` is not actually in the 7th position within the joint state array.
```
ASSERT_NE(curr_state.position[8], prev_state.position[8]);
```

ibrahiminfinite commented on moveit_ros/moveit_servo/include/moveit_servo/utils/datatypes.hpp:
```diff
@@ -36,34 +31,99 @@
+/*      Title       : datatypes.hpp
+ *      Project     : moveit_servo
+ *      Created     : 06/05/2023
+ *      Author      : V Mohammed Ibrahim
```
Oh sorry about that , I did take care to check that I included previous authors, some of it must have gotten lost during licence copy paste in between.

sea-bass reviewed (changes requested):
Finally got to a place where Servo ran on my robot and it does not work.

When I send joint jog commands, sometimes the wrong joint will move a little bit and then nothing happens ever again. I think there are a few more bugs left in the code.

I am sending the following joint jog command:

```
header:
  stamp:
    sec: 1690568309
    nanosec: 149086033
  frame_id: ''
joint_names:
- elbow_joint
displacements: []
velocities:
- 0.0
duration: 0.0
```

The joint delta is still coming out as a scalar, which is incorrect. Also, I don't think servo is correctly picking which joint is the one actually moving.

Have you tried publishing joint jog commands that only have a subset of the joint names specified, like in here?

I also think functions like `jointDeltaFromCommand` are not actually updating the `joint_position_deltas` variable unless the command type is invalid...

Then again, Twist commands also didn't work for me either. Servo is reading the command, but nothing is happening. For example:

```
header:
  stamp:
    sec: 1690569060
    nanosec: 963843584
  frame_id: manual_grasp_link
twist:
  linear:
    x: 1.0
    y: 0.0
    z: 0.0
  angular:
    x: 0.0
    y: 0.0
    z: 0.0
```

ibrahiminfinite commented:
For jointjog, the issue is that the conversion logic from ROS message to `ServoInput` assumes the joint jog commands contains a value for all joints. ( A wrong assumption), this can be easily fixed by updating the conversion logic. The joint jog command validation should also have a length check.

In case of of twist,  the `manual_grasp_link` frame_id is the issue, Servo expects the commands to be in the `servo_params.planning_frame`,

In both cases, the status of `Servo` will be (should be) `INVALID` and so it will not publish a message to the robot controller.

ibrahiminfinite commented:
> On MoveIt Studio we're using servo to move in the end effector frame, not the base frame. Requiring this is a huge regression.

While it works in the existing implementation due to the implicit frame conversion, the frame conversion formula used is not the proper one. https://github.com/ros-planning/moveit2/issues/2150

```C++
  // Convert the command to planning frame.
  tf2_ros::Buffer transform_buffer(demo_node->get_clock());
  tf2_ros::TransformListener transform_listener(transform_buffer);
  const auto command_to_planning_frame = transform_buffer.lookupTransform(
      servo_params.planning_frame, target_twist.frame_id, rclcpp::Time(0), rclcpp::Duration::from_seconds(2));
  Eigen::VectorXd transformed_twist = target_twist.velocities;
  tf2::doTransform(transformed_twist, transformed_twist, command_to_planning_frame);
  target_twist = TwistCommand{ servo_params.planning_frame, transformed_twist };
```
The above will do a transform, but it does not consider body frame vs space frame.

sea-bass commented:
> For jointjog, the issue is that the conversion logic from ROS message to `ServoInput` assumes the joint jog commands contains a value for all joints. ( A wrong assumption), this can be easily fixed by updating the conversion logic. The joint jog command validation should also have a length check.

I don't know. Because the jointjog command could specify any subset of joints, I think the validity of the command won't be based on length -- it will be based on whether the joints specified in the command are actually in the move group specified.

As far as checking lengths, I don't think runtime checks is what we need here. Instead, having some unit tests converting a ROS JointJog message (partially specified, fully specified, and with non-existing joints) and ensuring they are converted properly should do the job.

AndyZe commented:
I support Ibrahim on using the planning frame for all motions. We don't have any knowledge with tf2 on whether a frame is fixed in space or mobile, and those 2 cases require different equations. tf2 also doesn't give velocities. Having a separate utility library and requiring the user to make this decision seems like the best (only?) path forward.

sea-bass commented:
> I support Ibrahim on using the planning frame for all motions. We don't have any knowledge with tf2 on whether a frame is fixed in space or mobile, and those 2 cases require different equations. tf2 also doesn't give velocities. Having a separate utility library and requiring the user to make this decision seems like the best (only?) path forward.

Is there no other way that is less disruptive for everyone who wants to upgrade?

I'm growing concerned for my own project and for our users that our list of breaking changes now includes:
* Having to change a parameter name to define the end effector frame
* Replacing the start/stop services with a pause service
* Having to explicitly call a service to switch command types
* Having to pre-compute transforms for all commands not sent in the move group's planning frame

ibrahiminfinite commented:
> * Having to change a parameter name to define the end effector frame
> * Replacing the start/stop services with a pause service
> * Having to explicitly call a service to switch command types
> * Having to pre-compute transforms for all commands not sent in the move group's planning frame

1. If it will be a hassle, I am (hesitantly) okay with reverting it to old one.
2. The start/stop services can be put in as an alias with a depreciation message, so that they don't have to immediately make breaking changes to upgrade.
3. This cannot be avoided
4. The implicit transformation could be kept until we release the conversion utility library?

AndyZe commented:
Points 1-3 are pretty minor (especially #1). This is ROS Rolling where changes are OK. I definitely think the upcoming tutorial should have a section about upgrading - and thanks for beta testing the upgrade!

I guess I'm OK keeping the implicit transformation for now for pt 4 but it's going to have to change eventually. We fix bugs, not intentionally keep them.

sea-bass commented:
This all sounds like a reasonable plan to me -- thanks, guys!

sea-bass reviewed (commented):
Tested again.

## Speed
For some reason, whenever I ran the command publish rate wasn't going above 20 Hz. So any publish period higher than that (we were defaulting to 200 Hz in sim / 500 Hz on hardware) will simply not be met.

Could you on your end do a `ros2 topic hz` of the outgoing commands and compare your solution vs. the old servo for some high rate that can't be met? I'd like to see if it's my system or things truly got that much slower.

## Joint Jog
So, I could now get the robot to move with joint jog mode, but the joint indices were off by 1... so when I moved one joint, the next one would move. And then the first joint just wouldn't.

I suspect this might be the use of `getJointModelNames()` when it should be `getActiveJointModelNames()`. At least this change fixed it for me on this UR arm.

## Endpoint
Again I had to temporarily change my planning frame due to the limitation we spoke about.

I also realized that MoveIt Studio was not sending a `header.frame_id` in its twist command for some reason, so that invalidated my command, which is good. After hard-coding this for now, it worked!

https://github.com/ros-planning/moveit2/assets/4603398/2f52d56d-3f3b-450a-8f40-80d6e5e7d4bf

sea-bass commented:
Also, these tests are still not passing CI, so those need to work before we merge, of course.

```
    [moveit_servo_ros_integration_test-6] [  FAILED  ] ServoRosFixture.testJointJog
    [moveit_servo_ros_integration_test-6] [  FAILED  ] ServoRosFixture.testTwist
    [moveit_servo_ros_integration_test-6] [  FAILED  ] ServoRosFixture.testPose
```

ibrahiminfinite commented:
> For some reason, whenever I ran the command publish rate wasn't going above 20 Hz. So any publish period higher than that (we were defaulting to 200 Hz in sim / 500 Hz on hardware) will simply not be met.
> 
> Could you on your end do a `ros2 topic hz` of the outgoing commands and compare your solution vs. the old servo for some high rate that can't be met? I'd like to see if it's my system or things truly got that much slower.

This is the output of `ros2 topic hz /panda_arm_controller/joint_trajectory
` for pose command published at 500Hz
```
average rate: 498.530
	min: 0.000s max: 0.014s std dev: 0.00063s window: 5990
average rate: 498.559
	min: 0.000s max: 0.014s std dev: 0.00063s window: 6489
```
At 200Hz 
```
average rate: 199.997
	min: 0.000s max: 0.012s std dev: 0.00041s window: 602
average rate: 200.009
	min: 0.000s max: 0.012s std dev: 0.00036s window: 803
```
It should easily be possible to run Servo at 1000 Hz, the average processing time is only around 350 us, even with the latest changes that dynamically get the joint model data.

ibrahiminfinite commented:
Another thing I have been thinking about.
If a command takes time `Y` to process and the publish period is `X`.
For the loop to really publish every `X` seconds, shouldn't we sleep the loop for time `X - Y` rather than the whole `X` seconds ?

Don't know if I am missing something here.
What do you think @sea-bass  @AndyZe

ibrahiminfinite commented:
> Another thing I have been thinking about. If a command takes time `Y` to process and the publish period is `X`. For the loop to really publish every `X` seconds, shouldn't we sleep the loop for time `X - Y` rather than the whole `X` seconds ?
> 
> Don't know if I am missing something here. What do you think @sea-bass @AndyZe

I tried this at a command rate of 2000 Hz.

With  processing time compensation
```
average rate: 1956.193
	min: 0.000s max: 0.004s std dev: 0.00022s window: 10000
average rate: 1950.285
	min: 0.000s max: 0.004s std dev: 0.00021s window: 10000
average rate: 1953.905
	min: 0.000s max: 0.004s std dev: 0.00020s window: 10000
average rate: 1943.798
	min: 0.000s max: 0.004s std dev: 0.00021s window: 10000
```
without processing time compensation
```
average rate: 1889.547
	min: 0.000s max: 0.008s std dev: 0.00040s window: 10000
average rate: 1900.685
	min: 0.000s max: 0.008s std dev: 0.00037s window: 10000
average rate: 1899.729
	min: 0.000s max: 0.008s std dev: 0.00037s window: 10000
average rate: 1905.008
	min: 0.000s max: 0.007s std dev: 0.00037s window: 10000
```

AndyZe commented on moveit_ros/moveit_servo/src/servo.cpp:
```diff
@@ -33,74 +33,431 @@
+  return target_state;
+}
+
+const TwistCommand Servo::toPlanningFrame(const TwistCommand& command)
```
Need some error handling here to address issue #2150

sea-bass reviewed (changes requested):
Tested again with MoveIt Studio.

Things worked in mock hardware, except that after the stale command timeout occurs, the controller never stops publishing. It now sustains the rate even after you stop commanding the robot for a long time. The idea is the controller should sustain its rate until the `incoming_command_timeout` is reached, and then stop.

For a robot with perfect sensors, this is fine, but eventually the controller must stop publishing-- take, for example, Gazebo:

https://github.com/ros-planning/moveit2/assets/4603398/95853313-cf24-436d-a610-79390afb445c

---

Also, seems one of the unit tests for invalid vector is actually segfaulting?

ibrahiminfinite commented:
> Also, seems one of the unit tests for invalid vector is actually segfaulting?

Have to check what is happening in the CI, only the  "leaving singularity" test fails for me locally, no segfaults.

sea-bass commented on moveit_ros/moveit_servo/config/test_config_panda.yaml:
```diff
@@ -2,20 +2,20 @@
-publish_period: 0.034  # 1/Nominal publish rate [seconds]
+publish_period: 0.02  # 1/Nominal publish rate [seconds]
+
+incoming_command_timeout: 0.5 # seconds
```
Can this be brought back to its original 0.1 seconds for the timeout, or is there valid reason to increase it?

sea-bass commented on moveit_ros/moveit_servo/config/test_config_panda.yaml:
```diff
@@ -27,18 +27,21 @@ publish_joint_velocities: true
-
-## Stopping behaviour
-incoming_command_timeout:  0.1  # Stop servoing if X seconds elapse without a new command
+ee_frame: panda_link8  # The name of the end effector link, used to return the EE pose
```
I am still not sure why we had to change from `panda_hand` to `panda_link8` -- what was this again?

sea-bass commented on moveit_ros/moveit_servo/include/moveit_servo/servo_node.hpp:
```diff
@@ -69,6 +69,7 @@ class ServoNode
   ServoNode& operator=(ServoNode&) = delete;
 
+  // NOLINTNEXTLINE(readability-identifier-naming)
   rclcpp::node_interfaces::NodeBaseInterface::SharedPtr get_node_base_interface();
```
Rather than suppressing this, can we change it to `getNodeBaseInterface()`? Or is this something specifically inherited from rclcpp that we can't control?

sea-bass reviewed (commented):
This looks good to me!

I'll approve the `moveit_msgs` PR and left some very small things here.

sea-bass commented on moveit2.repos:
```diff
@@ -2,8 +2,8 @@ repositories:
     type: git
-    url: https://github.com/ros-planning/moveit_msgs.git
-    version: ros2
+    url: https://github.com/ibrahiminfinite/moveit_msgs.git
+    version: new_servo
```
This can now be updated as the PR is merged! 🎉

ibrahiminfinite commented:
Not sure if we need these https://github.com/ibrahiminfinite/moveit2/blob/eda31330572e27faf1e1cb4366cda3e6de843f7c/moveit_ros/moveit_servo/src/servo_node.cpp#L226
It floods the logs if the incoming command rate is low.

sea-bass commented:
> Not sure if we need these https://github.com/ibrahiminfinite/moveit2/blob/eda31330572e27faf1e1cb4366cda3e6de843f7c/moveit_ros/moveit_servo/src/servo_node.cpp#L226 It floods the logs if the incoming command rate is low.

Sure, feel free to get rid of it -- or put it as a DEBUG?
